trie/utils/verkle.go

```diff
@@ -0,0 +1,302 @@
+}
+
+func GetTreeKeyVersion(address []byte) []byte {
+	return GetTreeKey(address, zero, VersionLeafKey)
+}
+
+// Get does the,
+// but uses the cached point corresponding to the address part of the polynomial.
+// GetTreeKeyVersionWithEvaluatedAddress does the same thing as GetTreeKeyVersion,
+// but reuses a cached evaluation of the address part of the polynomial.
+func GetTreeKeyVersionWithEvaluatedAddress(addrp *verkle.Point) []byte {
+	return getTreeKeyWithEvaluatedAddess(addrp, zero, VersionLeafKey)
+}
+
+func GetTreeKeyBalance(address []byte) []byte {
+	return GetTreeKey(address, zero, BalanceLeafKey)
+}
+
+func GetTreeKeyNonce(address []byte) []byte {
+	return GetTreeKey(address, zero, NonceLeafKey)
+}
+
+func GetTreeKeyCodeKeccak(address []byte) []byte {
+	return GetTreeKey(address, zero, CodeKeccakLeafKey)
+}
+
+func GetTreeKeyCodeSize(address []byte) []byte {
+	return GetTreeKey(address, zero, CodeSizeLeafKey)
+}
+
+func GetTreeKeyCodeChunk(address []byte, chunk *uint256.Int) []byte {
+	chunkOffset := new(uint256.Int).Add(CodeOffset, chunk)
+	treeIndex := new(uint256.Int).Div(chunkOffset, VerkleNodeWidth)
+	subIndexMod := new(uint256.Int).Mod(chunkOffset, VerkleNodeWidth)
+	var subIndex byte
+	if len(subIndexMod) != 0 {
+		subIndex = byte(subIndexMod[0])
+	}
+	return GetTreeKey(address, treeIndex, subIndex)
+}
+
+// GetTreeKeyCodeChunkWithEvaluatedAddress does the same thing as GetTreeKeyCodeChunk,
+// but uses the cached point corresponding to the address part of the polynomial.
+func GetTreeKeyCodeChunkWithEvaluatedAddress(addressPoint *verkle.Point, chunk *uint256.Int) []byte {
+	chunkOffset := new(uint256.Int).Add(CodeOffset, chunk)
+	treeIndex := new(uint256.Int).Div(chunkOffset, VerkleNodeWidth)
+	subIndexMod := new(uint256.Int).Mod(chunkOffset, VerkleNodeWidth)
+	var subIndex byte
+	if len(subIndexMod) != 0 {
+		subIndex = byte(subIndexMod[0])
+	}
+	return getTreeKeyWithEvaluatedAddess(addressPoint, treeIndex, subIndex)
+}
+
+func GetTreeKeyStorageSlot(address []byte, storageKey *uint256.Int) []byte {
+	pos := storageKey.Clone()
+	if storageKey.Cmp(codeStorageDelta) < 0 {
+		pos.Add(HeaderStorageOffset, storageKey)
+	} else {
+		pos.Add(MainStorageOffset, storageKey)
+	}
+	treeIndex := new(uint256.Int).Div(pos, VerkleNodeWidth)
+
+	// calculate the sub_index, i.e. the index in the stem tree.
+	// Because the modulus is 256, it's the last byte of treeIndex
+	subIndexMod := new(uint256.Int).Mod(pos, VerkleNodeWidth)
+	var subIndex byte
+	if len(subIndexMod) != 0 {
+		// uint256 is broken into 4 little-endian quads,
+		// each with native endianness. Extract the least
+		// significant byte.
+		subIndex = byte(subIndexMod[0])
+	}
+	return GetTreeKey(address, treeIndex, subIndex)
+}
+
+func pointToHash(evaluated *verkle.Point, suffix byte) []byte {
+	// The output of Byte() is big engian for banderwagon. This
+	// introduces an imbalance in the tree, because hashes are
+	// elements of a 253-bit field. This means more than half the
```

Comment:
Something irrelevant, I don't get the point for using little endian here.

The Byte() return 353-bit hash, which has the last 3 bits as zero IIUC, why it will cause more than half the tree would be empty?